fe/fe-core/src/main/java/com/starrocks/common/util/concurrent/LockUtils.java

```diff
@@ -0,0 +1,172 @@
+
+        public void updateLastSlowLockLogTime(long newTime) {
+            lastSlowLockLogTime = newTime;
+        }
+
+        public long getLastSlowLockLogTime() {
+            return lastSlowLockLogTime;
+        }
+    }
+
+    private static void logSlowLockEventIfNeeded(long startMs, LockType type,
+                                                 long objectId, String objectName,
+                                                 JsonObject beforeLockInfo, JsonObject afterLockInfo,
+                                                 SlowLockLogStats slowLockLogStats) {
+        long endMs = TimeUnit.MILLISECONDS.convert(System.nanoTime(), TimeUnit.NANOSECONDS);
+        if (endMs - startMs > Config.slow_lock_threshold_ms &&
+                endMs > slowLockLogStats.getLastSlowLockLogTime() + Config.slow_lock_log_every_ms) {
+            slowLockLogStats.updateLastSlowLockLogTime(endMs);
+            JsonObject slowLockInfoJObj = new JsonObject();
+            slowLockInfoJObj.add("beforeLockInfo", beforeLockInfo);
+            slowLockInfoJObj.add("afterLockInfo", afterLockInfo);
+
+            // using json style instead of multi-line style, because for some logging collection
+            // system, it's hard to collect and format those log
+            LOG.warn("slow lock. type: {}, obj id: {}, obj name: {}, waited for: {}ms, info: {}",
+                    type, objectId, objectName, endMs - startMs, slowLockInfoJObj);
+        }
+    }
+
+    private static void logTryLockFailure(LockType type,
+                                          long timeout, TimeUnit unit,
+                                          long objectId, String objectName,
+                                          JsonObject beforeLockInfo, JsonObject afterLockInfo) {
+        JsonObject slowLockInfoJObj = new JsonObject();
+        slowLockInfoJObj.add("beforeLockInfo", beforeLockInfo);
+        slowLockInfoJObj.add("afterLockInfo", afterLockInfo);
+        LOG.warn("try lock failed. type: {}, timeout: {}, obj id: {}, obj name: {}, info: {}",
+                type, unit.toMillis(timeout), objectId, objectName, slowLockInfoJObj);
+    }
+
+    private static void lock(QueryableReentrantReadWriteLock rwLock,
+                             LockType type,
+                             long objectId, String objectName,
+                             SlowLockLogStats slowLockLogStats) {
+        long startMs = TimeUnit.MILLISECONDS.convert(System.nanoTime(), TimeUnit.NANOSECONDS);
+
+        JsonObject beforeLockInfo = getLockInfoWithOwnerStack(rwLock);
+        switch (type) {
+            case DB_READ_LOCK:
+                rwLock.sharedLock();
+                break;
+            case DB_WRITE_LOCK:
+                rwLock.exclusiveLock();
+                break;
+            default:
+                throw new UnsupportedOperationException("unknown lock type: " + type);
+        }
+        JsonObject afterLockInfo = getLockInfoWithCurrStack(rwLock);
```

Comment:
How about getting afterLockInfo only when printing slow log, I'm afraid it will block the thread while getting stack information.

Reply:
good idea , fixed